### ISSUE TITLE
False positive waiting on exported timeline semaphore that has had its win32 handle retrieved?

### ISSUE THREAD

**Issue by rygo6** [KhronosGroup/Vulkan-ValidationLayers#5295]
I seem to have run into the same bug highlighted in #1502 and #1999 that was then supposedly fixed in #2196 where it erroneously reports that you cannot wait on only a timeline semaphore submitted to a queue, but that you also need a fence.

It appears like #2196 did fix it except for when that timeline semaphore is also created with VkExportSemaphoreCreateInfo chained on pNext. Then you call vkGetSemaphoreWin32HandleKHR on the semaphore to get the WinNT handle.

So to be clear. This timeline semaphore works perfectly fine:
```
    const VkExportSemaphoreCreateInfo exportSemaphoreCreateInfo = {
            .sType = VK_STRUCTURE_TYPE_EXPORT_SEMAPHORE_CREATE_INFO,
            .handleTypes = externalHandleType,
    };
    const VkSemaphoreTypeCreateInfo timelineSemaphoreTypeCreateInfo = {
            .sType = VK_STRUCTURE_TYPE_SEMAPHORE_TYPE_CREATE_INFO,
            .pNext = &exportSemaphoreCreateInfo,
            .semaphoreType = VK_SEMAPHORE_TYPE_TIMELINE,
            .initialValue = 0,
    };
    const VkSemaphoreCreateInfo timelineSemaphoreCreateInfo = {
            .sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO,
            .pNext = &timelineSemaphoreTypeCreateInfo,
            .flags = 0,
    };
    FBR_VK_CHECK_RETURN(vkCreateSemaphore(pVulkan->device, &timelineSemaphoreCreateInfo, NULL, &pVulkan->timelineSemaphore));
```

If I also make the call to retrieve it's WinNT handle:
```
    VkSemaphoreGetWin32HandleInfoKHR semaphoreGetWin32HandleInfo = {
            .sType = VK_STRUCTURE_TYPE_SEMAPHORE_GET_WIN32_HANDLE_INFO_KHR,
            .pNext = VK_NULL_HANDLE,
            .handleType = externalHandleType,
            .semaphore = pVulkan->timelineSemaphore,
    };

    // TODO move to preloaded refs
    PFN_vkGetSemaphoreWin32HandleKHR getSemaphoreWin32HandleFunc = (PFN_vkGetSemaphoreWin32HandleKHR) vkGetInstanceProcAddr(pVulkan->instance, "vkGetSemaphoreWin32HandleKHR");
    if (getSemaphoreWin32HandleFunc == NULL) {
        FBR_LOG_DEBUG("Failed to get PFN_vkGetMemoryWin32HandleKHR!");
    }
    FBR_VK_CHECK_RETURN(getSemaphoreWin32HandleFunc(pVulkan->device, &semaphoreGetWin32HandleInfo, &pVulkan->externalTimelineSemaphore));
```

Only then does it start reporting the false positive on needing to wait on the fence. Triggering:
VUID-vkBeginCommandBuffer-commandBuffer-00045
VUID-vkBeginCommandBuffer-commandBuffer-00049
VUID-vkQueueSubmit2-commandBuffer-03875

**Comment by pborsutzki** [comment]
I am running into a similar problem, but upon destruction of a timeline semaphore after (successfully) waiting it on the host (`vkWaitSemaphores`). The validation layer reports a `VUID-vkDestroySemaphore-semaphore-01137`:

Validation Error: [ VUID-vkDestroySemaphore-semaphore-01137 ] Object 0: handle = 0x155d54390a0, name = vulkan device, type = VK_OBJECT_TYPE_DEVICE; \| MessageID = 0xa1569838 \| Cannot call vkDestroySemaphore on VkSemaphore 0xb977250000001328[] that is currently in use by a command buffer. The Vulkan spec states: All submitted batches that refer to semaphore must have completed execution (https://vulkan.lunarg.com/doc/view/1.3.236.0/windows/1.3-extensions/vkspec.html#VUID-vkDestroySemaphore-semaphore-01137)

I think this should not happen after successfully waiting for the semaphore. Especially as this only happens when also exporting the native windows handle through `vkGetSemaphoreWin32HandleKHR`. If that call is gone, the validation does not fail any more. With the call it fails independently of actually using the native windows handle.

Thanks for looking into this!

**Comment by artem-lunarg** [comment]
@pborsutzki thanks for confirmation of this issue. I'm currently working on external handles validation and will try to address this issue as part of that work.

**Comment by rygo6** [comment]
Is there any word on this issue? I still have a very terrible:
```
        // for some reason this fixes a bug with validation layers thinking the graphicsQueue hasnt finished
        // wait on timeline should be enough!!
        if (pVulkan->enableValidationLayers) {
            FBR_ACK_EXIT(vkQueueWaitIdle(pVulkan->graphicsQueue));
            FBR_ACK_EXIT(vkQueueWaitIdle(pVulkan->computeQueue));
        }
```
In my code :(

**Comment by artem-lunarg** [comment]
@rygo6 Sorry for delay. I'm starting with this issue and it is reproduceable.

**Comment by artem-lunarg** [comment]
@rygo6, @pborsutzki if you have the opportunity to test with the latest code please let me know if it fixes your issue.

**Comment by pborsutzki** [comment]
@artem-lunarg thank you for your efforts! Is there a simple way to test this without building the validation layers myself?

**Comment by juan-lunarg** [comment]
It is relatively simple to build

```
cmake -S . -B build -D UPDATE_DEPS=ON -D CMAKE_BUILD_TYPE=Release
cmake --build build/ --config Release
cmake --install build/ --prefix build/install --config Release # Binaries will be copied here
```

> @artem-lunarg thank you for your efforts! Is there a simple way to test this without building the validation layers myself?

I'm looking into making daily binaries via github. But for now you need to build it yourself.

**Comment by pborsutzki** [comment]
@artem-lunarg @juan-lunarg Thank you so much! I gave it a try, and it actually resolved the validation error I encountered with `vkDestroySemaphore` after exporting the native handle. I truly appreciate your help!

**Comment by artem-lunarg** [comment]
Happy to hear the issue was fixed. Going to close it.
@rygo6 if your issue persist let us know ideally though new issue with additional details, in that case it's a different issues then was fixed here.
